### PR TITLE
Fix dynamic max trials of RANSAC

### PR DESCRIPTION
**Pull request by hayatoikoma**
## Description

Due to numerical precision, `denom` sometimes becomes exactly `1.0` so that `np.log(denom)` becomes `0.0`. As the numerator is always negative, the function returns `-np.inf`. This PR fixes this issue.

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
Fix numerical precision error in RANSAC. In some cases, RANSAC was stopping at the first iteration.
```

### CODE REVIEW

File: skimage/measure/tests/test_fit.py
Comment (hayatoikoma) [inline skimage/measure/tests/test_fit.py:440]: On main, it returns `-np.inf`.

**Comment by hayatoikoma** [comment]
The CI seems to be failing due to numerical stability between different binaries? As I don't know the difference of the CIs, I am leaving it as it is.

File: skimage/measure/fit.py
Comment (lagru) [inline skimage/measure/fit.py:671]:
```suggestion
    denom = max(_EPSILON, 1 - inlier_ratio ** min_samples)
    denom = min(denom, 1 - _EPSILON)
```
What do you think about this? That way _EPSILON only has an impact on the results if it needs to. Might not matter at all though.

File: skimage/measure/tests/test_fit.py
Comment (lagru) [reply skimage/measure/tests/test_fit.py:440]: Can confirm. But why modify the existing test and not add another one precisely for the error case including a comment?

**Review by lagru — COMMENTED** [review]
Thanks. The [current test failures](https://github.com/scikit-image/scikit-image/actions/runs/6534156523/job/17740756432#step:5:625) look unrelated to this PR (imageio again). So ignore them for now. 

I've left a few minor suggestions, otherwise looks good.

File: skimage/measure/tests/test_fit.py
Comment (hayatoikoma) [reply skimage/measure/tests/test_fit.py]: With my first initial implementation, I thought that `min_samples = 5` would be redundant because the result for `min_samples=5` changes. But with your suggested change, the result doesn't change so that it makes sense to keep it. Just added it back with my new commit.

**Comment by hayatoikoma** [comment]
Thanks for the review! I just pushed the changes based on the suggestion.

File: skimage/measure/fit.py
Comment (lagru) [inline skimage/measure/fit.py:672]:
```suggestion
    # Avoid log(1) below turning into -inf
    denom = min(denom, 1 - _EPSILON)
```

**Review by lagru — APPROVED** [review]
Looks good, thanks. Feel welcome to tweak my suggested comment further.

**Review by lagru — APPROVED** [review]
Agreed, looks ready to go. :)

**Comment by hayatoikoma** [comment]
I just found that the function was a transplant from scikit-learn, and scikit-learn is handling the edge case in a slightly different way. I'm sure that it doesn't matter, but just FYI.

https://github.com/scikit-image/scikit-image/pull/6046

https://github.com/scikit-learn/scikit-learn/blob/2a2772a87b6c772dc3b8292bcffb990ce27515a8/sklearn/linear_model/_ransac.py#L23-L54

File: skimage/measure/fit.py
Comment (michaelbratsch) [reply skimage/measure/fit.py:672]: I think `denom = np.clip(1 - inlier_ratio**min_samples, _EPSILON, 1 - _EPSILON)`  is shorter an more expressive.

**Comment by michaelbratsch** [comment]
What is keeping this pull-request from being merged? I already opened another pull-request https://github.com/scikit-image/scikit-image/pull/7490 (without knowing that one exists) because the problem is quite annoying.

**Comment by lagru** [comment]
@michaelbratsch, sorry for that. We have the policy to only merge after two maintainers have reviewed and approved of a PR. So we are currently lacking a second approval. I am pinging @scikit-image/core to see if someone has the time take a look. It would indeed be sad to loose this bug fix to time! :pray:

File: skimage/measure/fit.py
Comment (lagru) [reply skimage/measure/fit.py]: Agreed. Though I think the 2nd and 3rd argument need to be switched. I took the liberty to add your suggestion in https://github.com/scikit-image/scikit-image/pull/7065/commits/f450f4ce4ccd2ba93520a8d08bd62be025bc68ec. I added you as a co-author but didn't know an email associated with your account so GitHub will not pick that up. Let me know if I should modify the message.

Comment (michaelbratsch) [reply skimage/measure/fit.py]: > Though I think the 2nd and 3rd argument need to be switched.

I do not understand what you mean. https://github.com/scikit-image/scikit-image/pull/7065#discussion_r1710976764 is the same as implemented in https://github.com/scikit-image/scikit-image/commit/f450f4ce4ccd2ba93520a8d08bd62be025bc68ec, right?

Additionally, I would also limit the nominator as in https://github.com/scikit-image/scikit-image/pull/7490/files. That way it is symmetric and (de-)nominator are both in (0, 1).

I added an email address to my GitHub account.

**Review by stefanv — APPROVED** [review]

File: skimage/measure/fit.py
Comment (lagru) [reply skimage/measure/fit.py]: Sorry, I was misreading your suggestion. I agree that guarding the nominator in a similar manner would be useful. Would you be willing to make a small PR for that? (Feel free to ping me!) That way you'll also get commit credits. :slightly_smiling_face:

Comment (michaelbratsch) [reply skimage/measure/fit.py]: Good suggestion. I opened a PR: https://github.com/scikit-image/scikit-image/pull/7496

**Pull request merged**
